File: Simulator.java

public class Simulator {
    final static int COOPERATION_REWARD = 2;
    final static int STEAL_REWARD = 3;
    final static int STEAL_PARTIAL_REWARD = 2;
    final static int STOLEN_PENALTY = -1;

    public static void run_round(Bot bot1,Bot bot2){
        //Takes 2 bots and subjects them to a match

        if (bot1.decide().equals("cooperate") && bot2.decide().equals("cooperate")){
            bot1.coins += COOPERATION_REWARD;
            bot2.coins += COOPERATION_REWARD;
            }
        else if (bot1.decide().equals("cheat") && bot2.decide().equals("cooperate")){
            if (bot2.coins >= STOLEN_PENALTY*-1){
                bot1.coins += STEAL_REWARD;
                bot2.coins += STOLEN_PENALTY;}
            else {
                bot1.coins += STEAL_PARTIAL_REWARD;
                }
            }
        else if (bot1.decide().equals("cooperate") && bot2.decide().equals("cheat")){
            if (bot1.coins >= STOLEN_PENALTY*-1){
                bot2.coins += STEAL_REWARD;
                bot1.coins += STOLEN_PENALTY;}
            else {
                bot2.coins += STEAL_PARTIAL_REWARD;
            }
        }
        else if (bot1.decide().equals("cheat") && bot2.decide().equals("cheat")){
            //no change in coins if both bots attempt to cheat
            }
        bot1.memory = bot2.decide();
        bot2.memory = bot1.decide();
        }
    public static void run_game(Bot bot1,Bot bot2,int rounds){
        //Takes 2 bots and subjects them to n rounds
        bot1.reset();
        bot2.reset();
        for (int i = 0; i <= rounds; i++){
            run_round(bot1,bot2);
        }
        bot1.bank+=bot1.coins;
        bot2.bank+=bot2.coins;
        report_game_results(bot1,bot2);
    }

    public static void report_game_results(Bot bot1, Bot bot2){
        //Reports winner of game after all rounds completed
        if (bot1.coins > bot2.coins) {
            System.out.println(bot1.toString()+ " wins the game against" + bot2.toString());
            }
        else if (bot1.coins < bot2.coins) {
            System.out.println(bot2.toString()+ " wins the game against" + bot1.toString());
        }
        if (bot1.coins == bot2.coins) {
            System.out.println("In the game between " + bot1.toString() + " and " + bot2.toString()+ " it's a tie");
        }
    }
}

File: simulator.py
from random import random
import itertools

def run_round(bot1,bot2):
    """Takes 2 bots and subjects them to a match"""
    if bot1.decide() == 'cooperate' and bot2.decide() == 'cooperate':
        bot1.coins+=2
        bot2.coins+=2
        bot1.memory = 'cooperate'
        bot2.memory = 'cooperate'
    elif bot1.decide() == 'cheat' and bot2.decide() == 'cooperate':
        bot1.coins+=3
        if bot2.coins > 0:
            bot2.coins-=1
        bot1.memory = 'cooperate'
        bot2.memory = 'cheat'
    elif bot1.decide() == 'cooperate' and bot2.decide() == 'cheat':
        if bot1.coins > 0:
            bot1.coins-=1
        bot2.coins+=3
        bot1.memory = 'cheat'
        bot2.memory = 'cooperate'
    elif bot1.decide() == 'cheat' and bot2.decide() == 'cheat':
        bot1.memory = 'cheat'
        bot2.memory = 'cheat'

def report_game(bot1,bot2):
    """Reports winner of game after rounds"""
    if bot1.coins > bot2.coins:
        return f"In match between {bot1} and {bot2}, {bot1} wins by {bot1.coins-bot2.coins}!"
    elif bot2.coins > bot1.coins:
        return f"In match between {bot1} and {bot2}, {bot2} wins by {bot2.coins-bot1.coins}!"
    else:
        return f"In match between {bot1} and {bot2}, its a tie! (Coins increased by {bot1.coins})"

def run_game(bot1,bot2,rounds):
    """Takes 2 bots and subjects them to n rounds, declare winner after each game"""
    bot1.reset()
    bot2.reset()
    for round in range(1,rounds+1):
        run_round(bot1,bot2)
    bot1.bank+=bot1.coins
    bot2.bank+=bot2.coins
    return report_game(bot1,bot2)

def run_generation(bots, rounds, track = False):
    """Run generation of games with n rounds each"""
    bot_pairs = list(itertools.combinations(bots, 2))

    for bot1, bot2 in bot_pairs:
        result = run_game (bot1, bot2, rounds=rounds)
        if track:
            print (result)
    print (rank_bots(bots))
    return bots

def create_bot(bot,n=1):
    """Create n bots and return as list"""
    return [bot() for i in range(1,n+1)]

def rank_bots(bots):
    """Rank bots based on amount of coins"""
    return sorted(bots, key=lambda x: x.bank, reverse = True)

def count_bots(bots):
    """Count bots in list, return dictionary """
    registry = {}
    for bot in bots:
        if type(bot).__name__ in registry:
            registry[type(bot).__name__]+= 1
        else:
            registry[type(bot).__name__] = 1
    return registry

def new_generation(bots,culled=5):
    """Create new generation of bots by removing bottom ranked and populating top ranked"""
    if culled != 0:
        bots = rank_bots(bots)
        top_bots = bots[:culled]
        middle_bots = bots[culled:-culled]
        bottom_bots = bots[-culled:]
        new_bots = [bot.spawn() for bot in top_bots]
        print ('Bots joing next generation: ',new_bots)
        bots =  top_bots + middle_bots + new_bots
    return bots


